### ISSUE TITLE
[Comb] extract narrowing could be more clever with and/or operations

### ISSUE THREAD

**Issue by lattner** [llvm/circt#1581]
With some recent patches, we are generating extract operations from `and` and `or` operations that have a constant RHS, we should know that we can forward through the LHS operand.

Concretely, the `combine_icmp_compare_known_bits0` example in Comb/canonicalization.mlir compiles to:

```
 hw.module @combine_icmp_compare_known_bits0(%thing: i4) -> (%a: i1) {
    %c0_i3 = hw.constant 0 : i3
    %c-3_i4 = hw.constant -3 : i4
    %0 = comb.and %thing, %c-3_i4 : i4
    %1 = comb.extract %0 from 2 : (i4) -> i2
    %2 = comb.extract %0 from 0 : (i4) -> i1
    %3 = comb.concat %1, %2 : (i2, i1) -> i3
    %4 = comb.icmp eq %3, %c0_i3 : i3
    hw.output %4 : i1
  }
```

but we should be able to eliminate the `and` operation, generating this instead:

```
 hw.module @combine_icmp_compare_known_bits0(%thing: i4) -> (%a: i1) {
    %c0_i3 = hw.constant 0 : i3
    %c-3_i4 = hw.constant -3 : i4
    %1 = comb.extract %thing from 2 : (i4) -> i2
    %2 = comb.extract %thing from 0 : (i4) -> i1
    %3 = comb.concat %1, %2 : (i2, i1) -> i3
    %4 = comb.icmp eq %3, %c0_i3 : i3
    hw.output %4 : i1
  }
```

since the and doesn't change the referenced bits.

**Comment by lattner** [comment]
@fyquah are you interested in this?

**Comment by lattner** [comment]
Implemented in b1ff1967
